### PR TITLE
NJ 193 - Add none of the above option to property tax

### PR DESCRIPTION
**Pull request by jachan**
## Link to pivotal/JIRA issue
- https://github.com/newjersey/affordability-pm/issues/193

## Is PM acceptance required? (delete one)
- Yes - don't merge until JIRA issue is accepted!

**Reminder**: merge main into this branch and get green tests before merging to main
## What was done?
- Add None of the Above checkbox to tenant and homeowner property tax eligibility
- None of the Above unchecks all other checkboxes when selected
- If none of the above is checked, and user checks another checkbox, none of the above becomes unchecked
- Moved NJ End to End tests to its own file, as it was getting very large (~400 lines out of a 1000 line test)

## How to test?
- Describe the testing approach taken to verify the changes, including:
  - Local Tests 
    - Select "Lauryn MFS" (this allows an additional checkbox to appear for spouse)
      - Navigate to Property Tax, select Tenant or Homeowner
      - Validate that None of the Above behavior is as expected
  - Added end-to-end tests

## Screenshots (for visual changes)
- Before
image

- After

https://github.com/user-attachments/assets/859c5b4c-b749-482f-8a94-baac5b914b57

### CODE REVIEW

File: spec/features/state_file/eligibility_spec.rb
Comment (jachan) [inline spec/features/state_file/eligibility_spec.rb:1]: Removed what appeared to be an empty test file.

File: spec/features/state_file/nj/complete_intake_spec.rb [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+      property_text = property.present? ? I18n.t("state_file.questions.nj_ineligible_property_tax.edit.#{property}") : nil
+      expect(page).to have_text strip_html_tags(I18n.t("state_file.questions.nj_ineligible_property_tax.edit.title_html", property: property_text)).strip
+      expect(page).to have_text strip_html_tags(I18n.t("state_file.questions.nj_ineligible_property_tax.edit.reason_html", reason: reason_text))
+      continue
+    end
+
+    def advance_to_property_tax_page(df_persona_name)
+      advance_to_start_of_intake(df_persona_name)
+      advance_county_and_municipality
+      if has_text? I18n.t("state_file.questions.nj_disabled_exemption.edit.title")
```
Comment (jachan) [inline spec/features/state_file/nj/complete_intake_spec.rb:165]: Made these conditional, as not all personas are asked about Disability (e.g. if Federal XML claims blindness, we take this exemption for them)

File: spec/features/state_file/nj/complete_intake_spec.rb [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+        advance_disabled_exemption
+      end
+      advance_veterans_exemption
+      if has_text? I18n.t("state_file.questions.nj_college_dependents_exemption.edit.title")
```
Comment (jachan) [inline spec/features/state_file/nj/complete_intake_spec.rb:169]: Similarly, only personas with dependents are shown this controller, so made this conditional as well

File: spec/features/state_file/nj/complete_intake_spec.rb [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+      it "handles property tax ineligible homeowner flow", required_schema: "nj" do
+        advance_to_property_tax_page("Zeus one dep")
+        choose_household_rent_own("homeowner")
+        select_homeowner_eligibility([])
+        expect_ineligible_page("on_home", "property_taxes")
+        expect_estimated_tax_page
+      end
+
+      it "handles homeowner none of the above checkbox", required_schema: "nj" do
```
Comment (jachan) [inline spec/features/state_file/nj/complete_intake_spec.rb:287]: New test added here for homeowner none of the above

File: spec/features/state_file/nj/complete_intake_spec.rb [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+
+      it "handles property tax ineligible tenant flow", required_schema: "nj" do
+        advance_to_property_tax_page("Zeus one dep")
+        choose_household_rent_own("tenant")
+        select_tenant_eligibility([])
+        expect_ineligible_page("on_rental", "property_taxes")
+        expect_estimated_tax_page
+      end
+
+      it "handles tenant none of the above checkbox", required_schema: "nj" do
```
Comment (jachan) [inline spec/features/state_file/nj/complete_intake_spec.rb:250]: New test added here for tenant none of the above

File: spec/features/state_file/nj/complete_intake_spec.rb
Comment (jachan) [inline spec/features/state_file/nj/complete_intake_spec.rb:1]: Commented on changes, as moving all NJ tests out to a separate file makes it difficult to see what was modified.

File: spec/features/state_file/nj/complete_intake_spec.rb [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+      case household_rent_own
+      when "homeowner", "tenant", "both", "neither"
+        choose I18n.t("state_file.questions.nj_household_rent_own.edit.#{household_rent_own}")
+      else
+        throw "not a valid choice"
+      end
+      continue
+    end
+
+    def select_homeowner_eligibility(checkboxes, continue_to_next: true)
```
Comment (jachan) [inline spec/features/state_file/nj/complete_intake_spec.rb:99]: Updated this helper function and the one below by adding `continue_to_next`

File: app/views/state_file/questions/nj_homeowner_eligibility/edit.html.erb [hunk 6]
```diff
@@ -43,12 +43,48 @@
...
+    document.addEventListener("DOMContentLoaded", function() {
+      let noneCheckbox = document.getElementById("state_file_nj_homeowner_eligibility_form_homeowner_none_of_the_above");
+      let checkboxes = [
+        document.getElementById("state_file_nj_homeowner_eligibility_form_homeowner_home_subject_to_property_taxes"),
+        document.getElementById("state_file_nj_homeowner_eligibility_form_homeowner_more_than_one_main_home_in_nj"),
+        document.getElementById("state_file_nj_homeowner_eligibility_form_homeowner_shared_ownership_not_spouse"),
+        document.getElementById("state_file_nj_homeowner_eligibility_form_homeowner_main_home_multi_unit"),
+        document.getElementById("state_file_nj_homeowner_eligibility_form_homeowner_main_home_multi_unit_max_four_one_commercial")
+      ]
+      let conditionalCheckbox = document.getElementById("state_file_nj_homeowner_eligibility_form_homeowner_same_home_spouse")
```
Comment (mmazanec22) [inline app/views/state_file/questions/nj_homeowner_eligibility/edit.html.erb:68]: [dust] I would prefer a more specific name for conditionalCheckbox, like maybe marriedFilingSeparatelyConditional or something?

File: app/views/state_file/questions/nj_tenant_eligibility/edit.html.erb [hunk 7]
```diff
@@ -65,4 +70,36 @@
 
     <%= f.continue %>
   <% end %>
-<% end %>
+  <script>
+    document.addEventListener("DOMContentLoaded", function() {
```
Comment (mmazanec22) [inline app/views/state_file/questions/nj_tenant_eligibility/edit.html.erb:74]: [pebble] Seeing that we're using the same logic twice, I wonder if it's worth adding the function passed to the event listener to `app/javascript. You could refactor it to work for both cases by passing in the IDs of all not-none checkboxes, for each id adding the checkbox to the checkbox array if they're not null, passing in the noneCheckbox ID, and then optional checkbox ids that should be hidden when none is checked

File: spec/features/state_file/nj/complete_intake_spec.rb
Comment (mmazanec22) [reply spec/features/state_file/nj/complete_intake_spec.rb:1]: Separating this out is a big improvement!

File: spec/features/state_file/nj/complete_intake_spec.rb (hunk 5 above)
Comment (mmazanec22) [reply spec/features/state_file/nj/complete_intake_spec.rb:99]: This pattern will be useful for a11y testing too!

**Review by mmazanec22 — APPROVED** [review]
Looks great! Non-blocking refactor suggestion

File: app/views/state_file/questions/nj_homeowner_eligibility/edit.html.erb (hunk 6 above)
Comment (jachan) [reply app/views/state_file/questions/nj_homeowner_eligibility/edit.html.erb:68]: I like that! It's clear what the checkbox is conditional on, which is nice. Renaming both homeowner and tenant eligibility with the new name.

File: app/views/state_file/questions/nj_tenant_eligibility/edit.html.erb (hunk 7 above)
Comment (mluedke2) [reply app/views/state_file/questions/nj_tenant_eligibility/edit.html.erb:74]: i like this suggestion too. we or other states might need this pattern in more places so it would be handy to reuse

File: spec/features/state_file/nj/complete_intake_spec.rb
Comment (mluedke2) [reply spec/features/state_file/nj/complete_intake_spec.rb:1]: i agree! i think over future PRs it would be useful to separate out even more, so that this file really is just the minimal steps for a "complete intake" and then we can get into the weeds on edge cases in other places... but that can be done later

**Review by mluedke2 — APPROVED** [review]
looks good! thanks for sorting tests too

File: spec/features/state_file/nj/complete_intake_spec.rb
Comment (mpidcock) [inline spec/features/state_file/nj/complete_intake_spec.rb:1]: [dust] I like that you separated this out! something to be cautious of is that these journey tests that walk through an entire intake are the longest running tests we have. we tend to use the "complete intake" spec to cover only the minimum flow to ensure all pages can load and in the right sequence. 

for less common/more complex pages you could consider splitting them out into a separate journey test, such as [this grocery credit page in idaho](https://github.com/codeforamerica/vita-min/blob/main/spec/features/state_file/id_grocery_credit_nested_followups_spec.rb).

**Review by mpidcock — APPROVED** [review]
Looks great!

**Review by mrotondo — APPROVED** [review]
IMO this looks good, and should have a followup PR to refactor the duplicated code into a separate JS file.

**Comment by jachan** [comment]
Rebased onto main, created separate followup ticket for refactor: https://github.com/newjersey/affordability-pm/issues/245. Planning to merge if all tests are green!
